File: Code Generation/responses/python/gemini/medium/apply-substitutions.py
from typing import List
import re

class Solution:
    def applySubstitutions(self, replacements: List[List[str]], text: str) -> str:
        # Convert replacements list to a dictionary for O(1) lookup
        mapping = {k: v for k, v in replacements}
        
        # Cache for memoization to handle nested substitutions efficiently
        memo = {}

        def resolve(val: str) -> str:
            """
            Recursively resolves all placeholders within a string.
            Since the problem guarantees no cyclic dependencies, this will terminate.
            """
            if val in memo:
                return memo[val]
            
            # Find all placeholders formatted as %VAR%
            # We use a while loop or regex sub to ensure we catch all nested instances
            result = val
            while '%' in result:
                # Find the first occurrence of a placeholder
                match = re.search(r'%([A-Z])%', result)
                if not match:
                    break
                
                placeholder = match.group(0) # e.g., "%A%"
                key = match.group(1)         # e.g., "A"
                
                # Recursively resolve the replacement value associated with the key
                replacement_resolved = resolve(mapping[key])
                
                # Replace the placeholder with its fully resolved value
                # We replace all occurrences of this specific key to be efficient
                result = result.replace(placeholder, replacement_resolved)
            
            memo[val] = result
            return result

        # The initial text is just another string that may contain placeholders
        return resolve(text)

File: Code Generation/responses/cpp/gemini/medium/apply-substitutions.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    /**
     * Resolves a key by recursively replacing any placeholders within its mapped value.
     * Since there are no cyclic dependencies, this process will eventually terminate.
     */
    string resolve(const string& key, unordered_map<string, string>& mapping, unordered_map<string, string>& memo) {
        if (memo.count(key)) {
            return memo[key];
        }

        string rawValue = mapping[key];
        string result = "";
        
        for (int i = 0; i < rawValue.length(); ++i) {
            if (rawValue[i] == '%') {
                // Find the closing '%'
                size_t nextPos = rawValue.find('%', i + 1);
                string innerKey = rawValue.substr(i + 1, nextPos - i - 1);
                result += resolve(innerKey, mapping, memo);
                i = nextPos; // Move index to the closing '%'
            } else {
                result += rawValue[i];
            }
        }

        return memo[key] = result;
    }

    string applySubstitutions(vector<vector<string>> replacements, string text) {
        unordered_map<string, string> mapping;
        unordered_map<string, string> memo;

        for (const auto& pair : replacements) {
            mapping[pair[0]] = pair[1];
        }

        string finalResult = "";
        for (int i = 0; i < text.length(); ++i) {
            if (text[i] == '%') {
                size_t nextPos = text.find('%', i + 1);
                string key = text.substr(i + 1, nextPos - i - 1);
                finalResult += resolve(key, mapping, memo);
                i = nextPos;
            } else {
                finalResult += text[i];
            }
        }

        return finalResult;
    }
};